contracts/mirror/src/contract.rs

```diff
@@ -0,0 +1,473 @@
+        config.source_channel = source_channel;
+    }
+    if let Some(ibc_timeout) = new_config.ibc_timeout {
+        attrs.push(attr("ibc_timeout", ibc_timeout.to_string()));
+        config.ibc_timeout = ibc_timeout;
+    }
+    if let Some(prefix) = new_config.prefix {
+        attrs.push(attr("prefix", &prefix));
+        config.prefix = prefix;
+    }
+    CONFIG.save(deps.storage, &config)?;
+    Ok(response("update_config", CONTRACT_NAME, attrs))
+}
+
+pub fn execute_change_return_type(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    info: MessageInfo,
+    id: u64,
+    return_type: ReturnType,
+) -> ContractResult<Response<NeutronMsg>> {
+    let mut bond = BONDS.load(deps.storage, id)?;
+    ensure!(
+        bond.state == BondState::Bonded,
+        ContractError::WrongBondState {
+            expected: BondState::Bonded.to_string(),
+            got: bond.state.to_string(),
+        }
+    );
+    let backup = bond.backup.clone().ok_or(ContractError::BackupIsNotSet)?;
+
+    ensure_eq!(info.sender, backup, ContractError::Unauthorized);
+    bond.return_type = return_type.clone();
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "change_return_type"),
+        attr("id", id.to_string()),
+        attr("return_type", return_type.to_string()),
+    ];
+    Ok(response("change_return_type", CONTRACT_NAME, attrs))
+}
+
+pub fn execute_update_bond(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    info: MessageInfo,
+    id: u64,
+    receiver: String,
+    backup: Option<String>,
+    return_type: ReturnType,
+) -> ContractResult<Response<NeutronMsg>> {
+    cw_ownable::assert_owner(deps.storage, &info.sender)?;
+    let mut bond = BONDS.load(deps.storage, id)?;
+    bond.receiver = receiver.clone();
+    bond.backup = backup
+        .clone()
+        .map(|a| deps.api.addr_validate(&a))
+        .transpose()?;
+    bond.return_type = return_type.clone();
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "update_bond"),
+        attr("id", id.to_string()),
+        attr("receiver", receiver),
+        attr("backup", backup.unwrap_or_default()),
+        attr("return_type", return_type.to_string()),
+    ];
+    Ok(response("update_bond_state", CONTRACT_NAME, attrs))
+}
+
+pub fn execute_bond(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    info: MessageInfo,
+    receiver: String,
+    r#ref: Option<String>,
+    backup: Option<String>,
+) -> ContractResult<Response<NeutronMsg>> {
+    let Config {
+        core_contract,
+        prefix,
+        ..
+    } = CONFIG.load(deps.storage)?;
+    let backup = backup
+        .as_ref()
+        .map(|addr| deps.api.addr_validate(addr))
+        .transpose()?;
+    bech32::decode(&receiver).map_err(|_| ContractError::WrongReceiverAddress)?;
+    ensure!(receiver.starts_with(&prefix), ContractError::InvalidPrefix);
+    let coin = cw_utils::one_coin(&info)?;
```

Comment:
I think we can change order of these operations and move this line into very beginning of this function